pkg/common/connectionstore/connection_recorder.go

```diff
@@ -0,0 +1,164 @@
+/*
+ *
+ * Copyright SecureKey Technologies Inc. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * /
+ *
+ */
+
+package connectionstore
+
+import (
+	"crypto"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+const (
+	stateNameCompleted = "completed"
+	myNSPrefix         = "my"
+	// TODO: https://github.com/hyperledger/aries-framework-go/issues/556 It will not be constant, this namespace
+	//  will need to be figured with verification key
+	theirNSPrefix    = "their"
+	errMsgInvalidKey = "invalid key"
+)
+
+// NewConnectionRecorder returns new connection recorder.
+// ConnectionRecorder is read-write connection store which provides
+// write features on top query features from ConnectionLookup
+func NewConnectionRecorder(p provider) (*ConnectionRecorder, error) {
+	lookup, err := NewConnectionLookup(p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection recorder : %w", err)
+	}
+
+	return &ConnectionRecorder{lookup}, nil
+}
+
+// ConnectionRecorder is read-write connection store
+type ConnectionRecorder struct {
+	*ConnectionLookup
+}
+
+// SaveInvitation saves invitation in permanent store for given key
+func (c *ConnectionRecorder) SaveInvitation(id string, invitation interface{}) error {
```

Comment:
I might need to better understand this function's signature?

Reply:
cyclic dependency issue, connection store cannot refer `didexchange.Invitation`

Reply:
It's weird to have a `SaveInvitation` in a generic connection recorder and still have the model in a specific protocol like did exchange.

Reply:
SaveInvitation happens during create-invitation of did-exchange,  which I believe should be out of connection recorder, it has nothing to do with the connections I believe.
It was moved in response to your comment on previous PR where you didn't like having `SaveInvitation` exported in DID Exchange Service for this purpose.

Created https://github.com/hyperledger/aries-framework-go/issues/1030 and will Add a TODO

Reply:
added TODO

Reply:
I think it might end up that Invitation need to be moved outside of DID Exchange as their own component.

Reply:
Updated #1030